src/Symfony/Bundle/FrameworkBundle/HttpClient/TraceableHttpClient.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\HttpClient;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpCache\Esi;
+use Symfony\Component\HttpKernel\HttpCache\HttpCache as BaseHttpCache;
+use Symfony\Component\HttpKernel\HttpCache\Store;
+use Symfony\Component\HttpKernel\KernelInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * AAA
+ *
+ * @author Jérémy Romey <jeremy@free-agent.fr>
+ */
+class TraceableHttpClient implements HttpClientInterface
+{
+    protected $httpClient;
+    protected $traces = [];
+
+    /**
+     * @param HttpClientInterface $httpClient A HttpClientInterface instance
+     */
+    public function __construct(HttpClientInterface $httpClient)
+    {
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $response = $this->httpClient->request($method, $url, $options);
+
+        $this->traces[] = [
+            'request' => [
+                'method' => $method,
+                'url' => $url,
+                'options' => $options,
+            ],
+            'response' => [
+                'statusCode' => $response->getStatusCode(),
+                'headers' => $response->getHeaders(false),
+            ],
```

Comment:
getting status code and headers here breaks response laziness and async behavior (since you have to wait for the response to be available, at least status code headers)

Reply:
Correct. Monitoring responses should be done by decorating the `on_progress` option instead.